crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1373,6 +1385,78 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Returns true if this type and `other` are gradual equivalent.
+    ///
+    /// > Two gradual types `A` and `B` are equivalent
+    /// > (that is, the same gradual type, not merely consistent with one another)
+    /// > if and only if all materializations of `A` are also materializations of `B`,
+    /// > and all materializations of `B` are also materializations of `A`.
+    /// >
+    /// > &mdash; [Summary of type relations]
+    ///
+    /// Note: `Todo != Todo`.
+    ///
+    /// This powers the `assert_type()` directive.
+    ///
+    /// [Summary of type relations]: https://typing.readthedocs.io/en/latest/spec/concepts.html#summary-of-type-relations
+    pub(crate) fn is_gradual_equivalent_to(self, db: &'db dyn Db, other: Type<'db>) -> bool {
+        let equivalent =
+            |(first, second): (&Type<'db>, &Type<'db>)| first.is_gradual_equivalent_to(db, *second);
+
+        if self == other && !matches!(self, Type::Todo(_)) {
+            return true;
+        }
+
+        match (self, other) {
+            (Type::Todo(_), Type::Todo(_)) => false,
+
+            (Type::Any | Type::Unknown, Type::Any | Type::Unknown) => true,
+
+            (Type::SubclassOf(first), Type::SubclassOf(second)) => {
+                match (first.subclass_of(), second.subclass_of()) {
+                    (ClassBase::Todo(_), ClassBase::Todo(_)) => false,
+                    (ClassBase::Any | ClassBase::Unknown, ClassBase::Any | ClassBase::Unknown) => {
+                        true
+                    }
+                    (ClassBase::Class(first), ClassBase::Class(second)) => first == second,
+                    _ => false,
+                }
+            }
+
+            (Type::Tuple(first), Type::Tuple(second)) => {
+                first.len(db) == second.len(db)
+                    && iter::zip(first.elements(db), second.elements(db)).all(equivalent)
+            }
+
+            (Type::Union(first), Type::Union(second)) => {
+                let first_elements = first.elements(db);
+                let second_elements = second.elements(db);
+
+                if first_elements.len() != second_elements.len() {
+                    return false;
+                }
+
+                let first_elements = first_elements.iter().collect::<OrderedTypeSet>();
+                let second_elements = second_elements.iter().collect::<OrderedTypeSet>();
+
+                iter::zip(first_elements, second_elements).all(equivalent)
+            }
+
+            (Type::Intersection(first), Type::Intersection(second)) => {
+                let first_positive = first.positive(db).iter().collect::<OrderedTypeSet>();
```

Comment:
Why do we need to collect the positive elements into another set when they are already set in `IntersectionType`? Could we instead compute the difference between `a` and `b` and then compare if those elements are equivalent?